I think there is a race condition with limit_concurrency

Environment:
- Latest version of SolidQueue
- Postgresql@16.3
- MacOS

My workflow that seems to reliably cause the issue:
- I have a workflow where I create approximately 10  records in couple of seconds.  
- This model has an after_commit block that enqueues 3 jobs via perform_later.
- The after_commit code is wrapped in a transaction (the 3 jobs really should be enqueued all-or-nothing), but the SolidJob error was occurring with and without the transactional wrapper

One of the Jobs fairly consistently generates an error on 'INSERT INTO "solid_queue_semaphores".
For each of the block of 10 or so creates, the offending jobs solid_queue_semephore "key" value is JobName/User.id, and all 10 of the commits will have the same key. 

What I believe I'm seeing via the logs is:
* N jobs are being enqueued in parallel (looks like N of 2)
* Because of some unfortunate random timing, these 2 enqueue requests are both attempting to insert at effectively the same time.
* Because they have the same solid_queue_semephore "key" value and the "key" has a unique constraint, one of the enqueues gets rejected with a 'ActiveRecord::StatementInvalid (PG::InFailedSqlTransaction: ERROR:  current transaction is aborted, commands ignored until end of transaction block' from deep within the bowels of Rails ActiveJob#enqueue.
* This is less than ideal ;-)

I haven't dug down into any of the SolidQueue code around enqueuing a job with limits_concurrency, but, at least with Postgres and it's version of MVCC, there is no way to preemptively detect a potential duplicate key prior to insert via a query.  It either has to be rescued or handled with an 'on conflict' clause (or the mySQL / sqlLite equivalent).

If my somewhat rushed conclusion is correct, then given limits_concurrency is a SolidQueue feature that enhances an enqueue request, it would be highly desirable for Solid to detect and retry internally.  Less desirable, but absolutely functional would be to raise an error that's specifically for this issue because the Job.perform_later is not the root cause, there isn't anything wrong with the database, you just got stepped on by combination of a database constraints and unfortunate timing, so the right answer is to resubmit job request it will likely work.

When I figure out how to get SolidQueue test environment working (I'm on a Mac and homebrew seems to require some kind of weird incantation to build a working version of mysql that works with docker), I'll see if I can develop a test for this specific case.

@rosa always appreciate the amount of support here!

> [This is what is being done](it would be highly desirable for Solid to detect and retry internally) and the basis of the concurrency limits. The unique key is precisely to avoid race conditions with multiple jobs that share a concurrency key being enqueued at the same time.

I understand the point of a unique key constraint, but as I had mentioned, it really depends on how you are using it.  I believe there are 2 distinct possibilities here: 

1. The key already exists, and a read or 'select for update' to avoid race conditions make detection easy.
2. The key does not exist...
     In postgresql, you can't 'select for update' a key that doesn't exist and can't proactively lock to avoid a race, you'd have to fallback to table locking if you are trying to 'read-verify' before writing. 
     Otherwise, you have to try to insert, then one threads will end up in a non-recoverable state at commit.

You are correct that my wrapping 3 jobs in a transaction is not making things easier.  That was simply a response to this randomly loosing one of my job requests.  A better way to handle this issue of random failures to enqueue is to change my code support a retry on a per job request.  

As for my ability to get a working test environment up and working.  bin/setup has always blown up on my laptop with the following.  And since I've never worked in a docker environment, this is befuddling to me:

```
bin/setup
creating Makefile

current directory: /Users/hms/.rvm/gems/ruby-3.2.2/gems/mysql2-0.5.4/ext/mysql2
make DESTDIR\= sitearchdir\=./.gem.20240513-34066-kirvj1 sitelibdir\=./.gem.20240513-34066-kirvj1 clean

current directory: /Users/hms/.rvm/gems/ruby-3.2.2/gems/mysql2-0.5.4/ext/mysql2
make DESTDIR\= sitearchdir\=./.gem.20240513-34066-kirvj1 sitelibdir\=./.gem.20240513-34066-kirvj1
compiling client.c
In file included from client.c:15:
./mysql_enc_name_to_ruby.h:43:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb_hash (str, len)
^
./mysql_enc_name_to_ruby.h:86:1: warning: a function definition without a prototype is deprecated in all versions of C and is not supported in C2x [-Wdeprecated-non-prototype]
mysql2_mysql_enc_name_to_rb (str, len)
^
client.c:1377:3: error: call to undeclared function 'mysql_ssl_set'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  mysql_ssl_set(wrapper->client,
  ^
client.c:1377:3: note: did you mean 'mysql_close'?
/opt/homebrew/opt/mysql/include/mysql/mysql.h:797:14: note: 'mysql_close' declared here
void STDCALL mysql_close(MYSQL *sock);
             ^
2 warnings and 1 error generated.
make: *** [client.o] Error 1

make failed, exit code 2
```

--- a/app/models/solid_queue/semaphore.rb
+++ b/app/models/solid_queue/semaphore.rb
@@ -66,11 +66,11 @@ module SolidQueue
         def check_limit_or_decrement = limit == 1 ? false : attempt_decrement
 
         def attempt_decrement
-          Semaphore.available.where(key: key).update_all(["value = value - 1, expires_at = ?", expires_at]) > 0
+          Semaphore.available.where(key: key).update_all([ "value = value - 1, expires_at = ?", expires_at ]) > 0
         end
 
         def attempt_increment
-          Semaphore.where(key: key, value: ...limit).update_all(["value = value + 1, expires_at = ?", expires_at]) > 0
+          Semaphore.where(key: key, value: ...limit).update_all([ "value = value + 1, expires_at = ?", expires_at ]) > 0
         end
 
         def key

--- a/app/models/solid_queue/semaphore.rb
+++ b/app/models/solid_queue/semaphore.rb
@@ -41,16 +41,30 @@ module SolidQueue
       end
 
       private
+
         attr_accessor :job
 
-        def attempt_creation
+        def attempt_creation_with_insert_on_conflict
+          results = Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
+
+          if results.length.zero?
+            limit == 1 ? false : attempt_decrement
+          else
+            true
+          end
+        end
+
+        def attempt_creation_with_create_and_exception_handling
           Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
           true
         rescue ActiveRecord::RecordNotUnique
-          if limit == 1 then false
-          else
-            attempt_decrement
-          end
+          limit == 1 ? false : attempt_decrement
+        end
+
+        if ActiveRecord::Base.connection.adapter_name == 'PostgreSQL'
+          alias attempt_creation attempt_creation_with_insert_on_conflict
+        else
+          alias attempt_creation attempt_creation_with_create_and_exception_handling
         end
 
         def attempt_decrement

--- a/app/models/solid_queue/semaphore.rb
+++ b/app/models/solid_queue/semaphore.rb
@@ -17,6 +17,17 @@ module SolidQueue
       def signal_all(jobs)
         Proxy.signal_all(jobs)
       end
+
+      # Requires a unique index on key
+      def create_unique_by(attributes)
+        if connection.supports_insert_conflict_target?
+          insert({ **attributes }, unique_by: :key).any?
+        else
+          create!(**attributes)
+        end
+      rescue ActiveRecord::RecordNotUnique
+        false
+      end
     end
 
     class Proxy
@@ -44,35 +55,22 @@ module SolidQueue
 
         attr_accessor :job
 
-        def attempt_creation_with_insert_on_conflict
-          results = Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
-
-          if results.length.zero?
-            limit == 1 ? false : attempt_decrement
-          else
+        def attempt_creation
+          if Semaphore.create_unique_by(key: key, value: limit - 1, expires_at: expires_at)
             true
+          else
+            check_limit_or_decrement
           end
         end
 
-        def attempt_creation_with_create_and_exception_handling
-          Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
-          true
-        rescue ActiveRecord::RecordNotUnique
-          limit == 1 ? false : attempt_decrement
-        end
-
-        if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
-          alias attempt_creation attempt_creation_with_insert_on_conflict
-        else
-          alias attempt_creation attempt_creation_with_create_and_exception_handling
-        end
+        def check_limit_or_decrement = limit == 1 ? false : attempt_decrement
 
         def attempt_decrement
-          Semaphore.available.where(key: key).update_all([ "value = value - 1, expires_at = ?", expires_at ]) > 0
+          Semaphore.available.where(key: key).update_all(["value = value - 1, expires_at = ?", expires_at]) > 0
         end
 
         def attempt_increment
-          Semaphore.where(key: key, value: ...limit).update_all([ "value = value + 1, expires_at = ?", expires_at ]) > 0
+          Semaphore.where(key: key, value: ...limit).update_all(["value = value + 1, expires_at = ?", expires_at]) > 0
         end
 
         def key

--- a/app/models/solid_queue/semaphore.rb
+++ b/app/models/solid_queue/semaphore.rb
@@ -61,7 +61,7 @@ module SolidQueue
           limit == 1 ? false : attempt_decrement
         end
 
-        if ActiveRecord::Base.connection.adapter_name == 'PostgreSQL'
+        if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
           alias attempt_creation attempt_creation_with_insert_on_conflict
         else
           alias attempt_creation attempt_creation_with_create_and_exception_handling

--- a/test/integration/concurrency_controls_test.rb
+++ b/test/integration/concurrency_controls_test.rb
@@ -83,7 +83,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # C would have started in the beginning, seeing the status empty, and would finish after
     # all other jobs, so it'll do the last update with only itself
-    assert_stored_sequence(@result, [ "C" ])
+    assert_stored_sequence(@result, ["C"])
   end
 
   test "run several jobs over the same record sequentially, with some of them failing" do
@@ -99,7 +99,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
     wait_for_jobs_to_finish_for(3.seconds)
     assert_equal 3, SolidQueue::FailedExecution.count
 
-    assert_stored_sequence @result, [ "B", "D", "F" ] + ("G".."K").to_a
+    assert_stored_sequence @result, ["B", "D", "F"] + ("G".."K").to_a
   end
 
   test "rely on dispatcher to unblock blocked executions with an available semaphore" do
@@ -133,7 +133,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # We can't ensure the order between B and C, because it depends on which worker wins when
     # unblocking, as one will try to unblock B and another C
-    assert_stored_sequence @result, ("A".."K").to_a, [ "A", "C", "B" ] + ("D".."K").to_a
+    assert_stored_sequence @result, ("A".."K").to_a, ["A", "C", "B"] + ("D".."K").to_a
   end
 
   test "rely on dispatcher to unblock blocked executions with an expired semaphore" do
@@ -165,7 +165,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # We can't ensure the order between B and C, because it depends on which worker wins when
     # unblocking, as one will try to unblock B and another C
-    assert_stored_sequence @result, ("A".."K").to_a, [ "A", "C", "B" ] + ("D".."K").to_a
+    assert_stored_sequence @result, ("A".."K").to_a, ["A", "C", "B"] + ("D".."K").to_a
   end
 
   test "don't block claimed executions that get released" do
@@ -183,42 +183,18 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
     assert job.reload.ready?
   end
 
-  if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
-    test "insert_with_unique_by_has_same_database_results_as_create!_with_exception_handling" do
-      key = "key", limit = 1, expires_at = 1.minute.from_now
+  test "verify transactions remain valid after Job creation conflicts via limits_concurrency" do
+    ActiveRecord::Base.transaction do
+      SequentialUpdateResultJob.perform_later(@result, name: "A", pause: 0.2.seconds)
+      SequentialUpdateResultJob.perform_later(@result, name: "B")
 
-      assert SolidQueue::Semaphore.count == 0
-
-      SolidQueue::Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
-      assert SolidQueue::Semaphore.count == 1
-
-      SolidQueue::Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
-      assert SolidQueue::Semaphore.count == 1
-
-      SolidQueue::Semaphore.delete_all
-      assert SolidQueue::Semaphore.count == 0
-
-      SolidQueue::Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
-      assert SolidQueue::Semaphore.count == 1
-
-      assert_raises ActiveRecord::RecordNotUnique do
-        SolidQueue::Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
+      begin
+        assert_equal 2, SolidQueue::Job.count
+        assert true, "Transaction state valid"
+      rescue ActiveRecord::StatementInvalid
+        assert false, "Transaction state unexpectedly invalid"
       end
     end
-  end
-
-  test "confirm correct version of attempt_creation by database adaptor" do
-    proxy = SolidQueue::Semaphore::Proxy.new(true)
-
-    aliased_method = proxy.method(:attempt_creation)
-
-    if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
-      original_method = proxy.method(:attempt_creation_with_insert_on_conflict)
-    else
-      original_method = proxy.method(:attempt_creation_with_create_and_exception_handling)
-    end
-
-    assert_equal original_method.name, aliased_method.original_name, "The alias maps to the correct original method"
   end
 
   private

--- a/test/integration/concurrency_controls_test.rb
+++ b/test/integration/concurrency_controls_test.rb
@@ -83,7 +83,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # C would have started in the beginning, seeing the status empty, and would finish after
     # all other jobs, so it'll do the last update with only itself
-    assert_stored_sequence(@result, ["C"])
+    assert_stored_sequence(@result, [ "C" ])
   end
 
   test "run several jobs over the same record sequentially, with some of them failing" do
@@ -99,7 +99,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
     wait_for_jobs_to_finish_for(3.seconds)
     assert_equal 3, SolidQueue::FailedExecution.count
 
-    assert_stored_sequence @result, ["B", "D", "F"] + ("G".."K").to_a
+    assert_stored_sequence @result, [ "B", "D", "F" ] + ("G".."K").to_a
   end
 
   test "rely on dispatcher to unblock blocked executions with an available semaphore" do
@@ -133,7 +133,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # We can't ensure the order between B and C, because it depends on which worker wins when
     # unblocking, as one will try to unblock B and another C
-    assert_stored_sequence @result, ("A".."K").to_a, ["A", "C", "B"] + ("D".."K").to_a
+    assert_stored_sequence @result, ("A".."K").to_a, [ "A", "C", "B" ] + ("D".."K").to_a
   end
 
   test "rely on dispatcher to unblock blocked executions with an expired semaphore" do
@@ -165,7 +165,7 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
 
     # We can't ensure the order between B and C, because it depends on which worker wins when
     # unblocking, as one will try to unblock B and another C
-    assert_stored_sequence @result, ("A".."K").to_a, ["A", "C", "B"] + ("D".."K").to_a
+    assert_stored_sequence @result, ("A".."K").to_a, [ "A", "C", "B" ] + ("D".."K").to_a
   end
 
   test "don't block claimed executions that get released" do

--- a/test/integration/concurrency_controls_test.rb
+++ b/test/integration/concurrency_controls_test.rb
@@ -183,7 +183,46 @@ class ConcurrencyControlsTest < ActiveSupport::TestCase
     assert job.reload.ready?
   end
 
+  if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
+    test "insert_with_unique_by_has_same_database_results_as_create!_with_exception_handling" do
+      key = "key", limit = 1, expires_at = 1.minute.from_now
+
+      assert SolidQueue::Semaphore.count == 0
+
+      SolidQueue::Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
+      assert SolidQueue::Semaphore.count == 1
+
+      SolidQueue::Semaphore.insert({ key: key, value: limit - 1, expires_at: expires_at }, unique_by: :key)
+      assert SolidQueue::Semaphore.count == 1
+
+      SolidQueue::Semaphore.delete_all
+      assert SolidQueue::Semaphore.count == 0
+
+      SolidQueue::Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
+      assert SolidQueue::Semaphore.count == 1
+
+      assert_raises ActiveRecord::RecordNotUnique do
+        SolidQueue::Semaphore.create!(key: key, value: limit - 1, expires_at: expires_at)
+      end
+    end
+  end
+
+  test "confirm correct version of attempt_creation by database adaptor" do
+    proxy = SolidQueue::Semaphore::Proxy.new(true)
+
+    aliased_method = proxy.method(:attempt_creation)
+
+    if ActiveRecord::Base.connection.adapter_name == "PostgreSQL"
+      original_method = proxy.method(:attempt_creation_with_insert_on_conflict)
+    else
+      original_method = proxy.method(:attempt_creation_with_create_and_exception_handling)
+    end
+
+    assert_equal original_method.name, aliased_method.original_name, "The alias maps to the correct original method"
+  end
+
   private
+
     def assert_stored_sequence(result, *sequences)
       expected = sequences.map { |sequence| "seq: " + sequence.map { |name| "s#{name}c#{name}" }.join }
       skip_active_record_query_cache do

